File: services/fx.py
from services.db import db


def get_fx_rate(from_ccy: str, to_ccy: str) -> float:
    if from_ccy == to_ccy:
        return 1.0
    for rate in db.get_fx_rates():
        if rate["baseCurrency"] == from_ccy and rate["quoteCurrency"] == to_ccy:
            return float(rate["rate"])
        if rate["baseCurrency"] == to_ccy and rate["quoteCurrency"] == from_ccy:
            r = float(rate["rate"])
            return 1.0 / r if r != 0 else 1.0
    return 1.0


def convert_amount(amount: float, from_ccy: str, to_ccy: str) -> float:
    return amount * get_fx_rate(from_ccy, to_ccy)

File: lib/services/fx.ts
import { db } from "@/lib/db";
import { BaseCurrency } from "@/lib/types";

export const convertAmount = (
  amount: number,
  fromCurrency: string,
  toCurrency: BaseCurrency
) => {
  if (fromCurrency === toCurrency) {
    return amount;
  }

  const rates = db.getFxRates();
  const direct = rates.find(
    (item) =>
      item.baseCurrency.toUpperCase() === fromCurrency.toUpperCase() &&
      item.quoteCurrency.toUpperCase() === toCurrency.toUpperCase()
  );

  if (direct) {
    return amount * direct.rate;
  }

  const inverse = rates.find(
    (item) =>
      item.baseCurrency.toUpperCase() === toCurrency.toUpperCase() &&
      item.quoteCurrency.toUpperCase() === fromCurrency.toUpperCase()
  );

  if (inverse) {
    return amount / inverse.rate;
  }

  return amount;
};

export const getFxOverview = () => db.getFxRates();
